#### Q11. 
$$J: R^3 \rightarrow R^2 $$
인 다음 벡터 함수의 야코비안 행렬을 구하시오.
$$ f(x) = (3.9x_1+x^3_2 - 2.3x^2_3, -2x_1x_2x^2_3)^T$$

In [10]:
# jacobian matrix
import numpy as np

def f(x1, x2, x3):
    f = (3.9*x1 + x2^3 - 2.3*x3^2 - 2*x1*x2*x3^2).T
    return f

def jacobian(x1, x2, x3):
    jacobian = np.array([
        [3.9,  -2*(x2)*(x3**2)], 
        [3(x2**2), -2*(x1)*(x3**2)], 
        [-4.6*x3 - 4*x1*x2*x3]
        ])
    return jacobian

[Solution] \
$\begin{bmatrix} 3.9&-2x_2x_3^3\\3x_2^2&-2x_1x_3^3\\ -4.6x_3&-4x_1x_2x_3\end{bmatrix}$

 #### Q12. 
 
 다음 함수에 대해 답하시오

$$
f(x)=2(\frac{1}{4}(1-2x)^2-1)^3 - 3(\frac{1}{4}(1-2x)^2-1)^2 - 3\\
$$

(1) $i\`(x)$와 $h\`(x)$를 쓰시오. \
(2) 연쇄법칙을 이용하여 $f\`(x)$를 구하시오. \
(3) $f\`(0), f\`(2.1)$을 계산하시오.

In [13]:
def f(x):
    return 2*x**3 - 3*x**2 + 1

def h(x):
    return ((x**2)/4)-1

def i(x):
    return (-2*x + 1)

def derivative_f(x):
    return 6*x**2 - 6*x

def derivative_h(x):
    return (x/2)

def derivative_i(x):
    return -2

def chain_rule(x):
    return derivative_f(h(i(x))) * derivative_h(i(x)) * derivative_i(x)

print("f'(0): {}".format(chain_rule(0.0)))
print("f'(2.1): {}".format(chain_rule(2.1))) 

f'(0): -7.875
f'(2.1): 16.773120000000016


[Soultion]
$$
i(x)=(1-2x)\\
h(x) = (\frac{1}{4}x^2-1)
$$

then, 
$$
f(x)=2x^3 - 3x^2 - 3
$$

$$
\frac{dy}{dx} = \frac{dy} {\cancel{dt}}  * \frac{\cancel{dt}}{\cancel{du}} * \frac{\cancel{du}}{dx}\\
$$


$$
= f'(t) * h'(u) * i'(x)\\
$$


$$
f'(x)=6x(x - 1) \\
h'(u)=\frac{1}{2}u\\
i'(v)=-2\\
$$


 #### Q13.
12번 문제의 f(x)에서 난수를 생성하여 초깃값 $x_0=2.1$을 얻었다고 가정하자.  
식 2.58을 연속적으로 적용하여 얻는 점 $x_1,x_2,x_3$를 구하시오.   
이때 학습률 p=0.1을 사용하시오. 점점 낮은곳을 향해 이동하는지 확인하시오.  

In [27]:
import plotly.graph_objects as go
from plotly.subplots import make_subplots

def f(x):
    return 2*x**3 - 3*x**2 + 1

def h(x):
    return ((x**2)/4)-1

def i(x):
    return (-2*x + 1)

def derivative_f(x):
    return 6*x**2 - 6*x

def derivative_h(x):
    return (x/2)

def derivative_i(x):
    return -2

def chain_rule(x):
    return derivative_f(h(i(x))) * derivative_h(i(x)) * derivative_i(x)

def f(x):
    return 2*x**3 - 3*x**2 + 1

def h(x):
    return ((x**2)/4)-1

def i(x):
    return (-2*x + 1)

def derivative_f(x):
    return 6*x**2 - 6*x

def derivative_h(x):
    return (x/2)

def derivative_i(x):
    return -2

def chain_rule(x):
    return derivative_f(h(i(x))) * derivative_h(i(x)) * derivative_i(x)

theta = 2.1
theta2 = 2.1
p = 0.1 
p2 = 0.001

fig = go.Figure()

fig = make_subplots(
    rows=2, cols=1,
    subplot_titles=("p = 0.1", "p = 0.001"),
    shared_xaxes=True,
    shared_yaxes=True,
    )

theta_list1 = []
theta_list2 = []

for g in range (100):
    theta_list1.append(theta)
    theta_list2.append(theta2)
    # print("x{} 대입 시 기울기: {}".format(g, chain_rule(theta))) 
    theta = theta - (chain_rule(theta))*p
    theta2 = theta2 - (chain_rule(theta2))*p2

fig.add_trace(
        go.Scatter(y=theta_list1, mode="lines+markers"),
        row=1, col=1
    )

fig.add_trace(
        go.Scatter(y=theta_list2, mode="lines+markers"),
        row=2, col=1
    )

fig.show()

####  Q 14. 다음 함수의 헤시안 행렬을 구하시오.
$$fx = f(x1, x2, x3) = x_1^2 + x_1x_2 (-4 + 4x_2^2)x_2^2 + x_1x_2x_3 + x_3^2$$

In [21]:
def f(x1, x2, x3):
    f = x1^2 + x1*x2*(-4 + 4*x2^2)*x2^2 + x1*x2*x3 + x3^2
    return f

def hessian(x1, x2, x3):
    hessian = np.array([
        [2, 1+x3, x2],
        [1+x3, -8+48*x2, x1],
        [x2, x1, 2]
    ])
    return hessian

[Solution] 
  
$\begin{bmatrix} 2&1+x_3&x_2\\1+x_3&48x_2-8&x_1\\ x_2 & x_1 & 2\end{bmatrix}$

#### Q15. 다음 함수에 대해 답하시오.
$$f(x_1,x_2) = 2x_1^2 + 3x_1x_2 + 2x_2^2 - 4x_1 + 2x_2 - 24$$

##### (1) 최솟값과 최소점을 분석적으로 구하시오.
##### (2) 난수를 생성하여 초기값 $x_0=(1.0, 0.9)^T$를 얻었다고 가정하고, 식(2.58)을 연속적으로 적용하여 얻는 점 $x_1, x_2, x_3$를 구하시오. 이 때 학습률 p=0.1을 사용하시오. (1)에서 구한 최소점을 향해 이동하는지 확인하시오,


In [11]:
import numpy as np
import plotly.graph_objects as go

p = 0.1

def fx(x):
    f = 2*x[0]**2 + 3*x[0]*x[1] + 2*x[1]**2 - 4*x[0] + 2*x[1] + 24
    return f

def gradient(x):
    a = 4*x[0] + 3*x[1] - 4
    b = 3*x[0] + 4*x[1] + 2
    return a, b

x_0 = np.array([1.0, 0.9])
gradient_x0 = np.array(gradient(x_0))

x_list = [x_0]
y_list = []
fig = go.Figure()

for i in range(1000):
    grad = np.array(gradient(x_list[i]))
    x_list.append(x_list[i] - p*grad)
    y_list.append(fx(x_list[i]))

    fig.add_trace(
        go.Scatter(
            # x=i, 
            y=y_list, 
            mode='lines+markers')
        )

    #ealry stopping
    if i > 0 and (y_list[i] - y_list[i-1] == 0):
        print("early stopping")
        print("step: {}".format(i))
        break

fig.show()

for i in range(4):
    print("Step: {i}".format(i=i))
    print("x1: {}".format(x_list[i][0]))
    print("x2: {}".format(x_list[i][1]))
    print("f(x1, x2): {}".format(fx(x_list[i])))


early stopping
step: 161


Step: 0
x1: 1.0
x2: 0.9
f(x1, x2): 28.12
Step: 1
x1: 0.73
x2: 0.040000000000000036
f(x1, x2): 22.3166
Step: 2
x1: 0.826
x2: -0.39499999999999996
f(x1, x2): 20.603792
Step: 3
x1: 1.0141
x2: -0.6848
f(x1, x2): 19.48533266


[Solution]   

(1) 최소점과 최솟값을 분석적으로 구하시오.  

$$ \frac {df(x_1, x_2)}{dx_1} = 4x_1 + 3x_2 - 4  $$
$$ \frac {df(x_1, x_2)}{dx_2} = 3x_1 + 4x_2 + 2  $$


최소점은 아래의 두식을 만족하는 $x_1$와 $x_2$이다.


$$ \frac {df(x_1, x_2)}{dx_1} = 0  $$
$$ \frac {df(x_1, x_2)}{dx_2} = 0  $$

위의 두식에 대하여 아래와 같은 연립방정식을 풀면 
$$
\begin{cases} 
4x_1 + 3x_2 - 4 = 0  \\
3x_1 + 4x_2 + 2 = 0
\end{cases}
$$
최소점은 아래와 같다. 

$$ x_1 = \frac{22}{7} = 3.142857142857143 $$
$$ x_2 = -\frac{20}{7} = -2.857142857142857 $$

위의 최소점을 바탕으로 최솟값은

$$f(x_1,x_2) = 2x_1^2 + 3x_1x_2 + 2x_2^2 - 4x_1 + 2x_2 - 24$$

$$ f(\frac{22}{7}, -\frac{20}{7}) = 2(\frac{22}{7})^2 + 3(\frac{22}{7})(-\frac{20}{7}) + 2(-\frac{20}{7})^2 - 4(\frac{22}{7}) + 2(-\frac{20}{7}) -24 \\
\\
= -33.14285714285714 $$

#### Q16.
##### [예제 2-11]의 테일러 급수에서 delta = 0.2, 0.1, 0.01, 0.001 일 때 계산하고, delta가 정확도에 미치는 영향을 설명하시오.
$$f(x) = \frac{1}{2}x^2$$

Taylor
$$f(x + \Delta) \simeq f(x)+f'(x)\Delta$$

In [4]:
# Q16.Taylor series

def fx(x):
    return 1/2 * x**2

def derivative_fx(x):
    return x

def taylor(x):
    return fx(x0) + derivative_fx(x0) * (x - x0)

delta = [0.2, 0.1, 0.01, 0.001]

# print("x0 = 1")

for i in delta:
    x0 = 1
    print("delta = ", i)
    print("x = ", x0 + i)
    print("f(x) = ", fx(x0 + i))
    print("taylor = ", taylor(x0 + i))
    print("error = ", abs(fx(x0 + i) - taylor(x0 + i)))
    print("=====")

delta =  0.2
x =  1.2
f(x) =  0.72
taylor =  0.7
error =  0.020000000000000018
=====
delta =  0.1
x =  1.1
f(x) =  0.6050000000000001
taylor =  0.6000000000000001
error =  0.0050000000000000044
=====
delta =  0.01
x =  1.01
f(x) =  0.51005
taylor =  0.51
error =  4.999999999999449e-05
=====
delta =  0.001
x =  1.001
f(x) =  0.5010004999999998
taylor =  0.5009999999999999
error =  4.999999999588667e-07
=====


[Solution]  
x = 1.0이라는 점에서 함수값 f(1.0) = 0.5,
그 점에서의 미분값 f'(1.0) = 1.0.

따라서 다음과 같은 답이 도출된다.  
f(1.0 + 0.2) = f(1.2) \
= f(1.0) + (f'(1.0)*0.2) \
=  0.5 + (1.0. * 0.2) = 0.7  


f(1.0 + 0.1) = 0.6  
f(1.0 + 0.01) = 0.51  
f(1.0 + 0.001) = 0.501

#### Q17.
##### [알고리즘 2-5]를 그 아래 코드로 바꾸어 쓰면, 한 번도 뽑히지 않는 샘플이 있을 수 있다. 그래도 문제가 없나? n개의 샘플이 있고, 뽑는 일을 T번 반복했을 때, 한번도 뽑히지 않은 샘플이 k개일 확률을 쓰시오.

In [3]:
from math import factorial
import numpy as np
import plotly.graph_objects as go

def bin_dist(n, k, t):
    ncr = factorial(n) / (factorial(k) * factorial(n - k))
    pd = ncr * ((n-k)/n)**t
    return pd

case = 5 # 샘플 수 
iter_num = 5 # 시행 횟수
k = 5 # 안뽑히는 개수

fig = go.Figure()

for j in range(k+1):
    pd = np.array(bin_dist(case, j, iter_num)) 
    # print(pd)

    fig.add_trace(
        go.Bar(
            x = [j],
            y = [pd],
            name = f'k = {j}',
            marker_line_color='rgb(8,48,107)',
            marker_line_width=1)
    )

fig.show()